packages/enzyme/src/ShallowWrapper.js

```diff
@@ -310,78 +311,60 @@ class ShallowWrapper {
         if (context) {
           this[OPTIONS] = { ...this[OPTIONS], context: nextContext };
         }
         this[RENDERER].batchedUpdates(() => {
+          // When shouldComponentUpdate returns false we shouldn't call componentDidUpdate.
+          // so we spy shouldComponentUpdate to get the result.
           let shouldRender = true;
-          // dirty hack:
-          // make sure that componentWillReceiveProps is called before shouldComponentUpdate
-          let originalComponentWillReceiveProps;
+          let spy;
           if (
             !this[OPTIONS].disableLifecycleMethods &&
             instance &&
-            typeof instance.componentWillReceiveProps === 'function'
+            typeof instance.shouldComponentUpdate === 'function'
           ) {
-            instance.componentWillReceiveProps(nextProps, nextContext);
-            originalComponentWillReceiveProps = instance.componentWillReceiveProps;
-            instance.componentWillReceiveProps = () => {};
+            spy = spyMethod(instance, 'shouldComponentUpdate');
+          }
+          if (props) this[UNRENDERED] = cloneElement(adapter, this[UNRENDERED], props);
+          this[RENDERER].render(this[UNRENDERED], nextContext);
+          if (spy) {
+            shouldRender = spy.getLastReturnValue();
+            spy.restore();
           }
-          // dirty hack: avoid calling shouldComponentUpdate twice
-          let originalShouldComponentUpdate;
           if (
+            shouldRender &&
             !this[OPTIONS].disableLifecycleMethods &&
-            instance &&
-            typeof instance.shouldComponentUpdate === 'function'
+            instance
           ) {
-            shouldRender = instance.shouldComponentUpdate(nextProps, state, nextContext);
-            originalShouldComponentUpdate = instance.shouldComponentUpdate;
-          }
-          if (shouldRender) {
-            if (props) this[UNRENDERED] = cloneElement(adapter, this[UNRENDERED], props);
-            if (originalShouldComponentUpdate) {
-              instance.shouldComponentUpdate = () => true;
-            }
-
-            this[RENDERER].render(this[UNRENDERED], nextContext);
-
-            if (originalShouldComponentUpdate) {
-              instance.shouldComponentUpdate = originalShouldComponentUpdate;
-            }
             const lifecycles = getAdapterLifecycles(adapter);
+            // getSnapshotBeforeUpdate
+            let snapshot;
             if (
-              !this[OPTIONS].disableLifecycleMethods &&
-              instance
+              lifecycles.getSnapshotBeforeUpdate
+              && typeof instance.getSnapshotBeforeUpdate === 'function'
             ) {
-              if (
-                lifecycles.getSnapshotBeforeUpdate
-                && typeof instance.getSnapshotBeforeUpdate === 'function'
-              ) {
-                const snapshot = instance.getSnapshotBeforeUpdate(prevProps, state);
```

Comment:
i have a mild preference for splitting the "snapshot" and "context" paths entirely, instead of trying to interleave them with `let snapshot`